src/bin/pg_autoctl/fsm_transition.c

```diff
@@ -1082,3 +1112,233 @@ fsm_promote_standby(Keeper *keeper)
+	 * setup where there is no primary_conninfo and otherwise use the same
+	 * parameters as for streaming replication.
+	 */
+	NodeAddress upstreamNode = { 0 };
+
+	if (!standby_init_replication_source(postgres,
+										 &upstreamNode,
+										 PG_AUTOCTL_REPLICA_USERNAME,
+										 config->replication_password,
+										 config->replication_slot_name,
+										 config->maximum_backup_rate,
+										 config->backupDirectory,
+										 NULL, /* no targetLSN */
+										 config->pgSetup.ssl,
+										 keeper->state.current_node_id))
+	{
+		/* can't happen at the moment */
+		return false;
+	}
+
+	log_info("Restarting standby node to disconnect replication "
+			 "from failed primary node, to prepare failover");
+
+	if (!standby_restart_with_current_replication_source(postgres))
+	{
+		log_error("Failed to disconnect from failed primary node, "
+				  "see above for details");
+
+		return false;
+	}
+
+	/*
+	 * Fetch most recent metadata, that will be sent in the next node_active()
+	 * call.
+	 */
+	if (!pgsql_get_postgres_metadata(pgsql,
+									 &pgSetup->is_in_recovery,
+									 postgres->pgsrSyncState,
+									 postgres->currentLSN,
+									 &(postgres->postgresSetup.control)))
+	{
+		log_error("Failed to update the local Postgres metadata");
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * When the selected failover candidate does not have the latest received WAL,
+ * it fetches them from another standby, the one in WAIT_CASCADE state.
+ */
+bool
+fsm_fast_forward(Keeper *keeper)
+{
+	KeeperConfig *config = &(keeper->config);
+	Monitor *monitor = &(keeper->monitor);
+	LocalPostgresServer *postgres = &(keeper->postgres);
+	PostgresSetup *pgSetup = &(postgres->postgresSetup);
+	int groupId = keeper->state.current_group;
+
+	NodeAddress upstreamNode = { 0 };
+	ReplicationSource replicationSource = { 0 };
+
+	char slotName[MAXCONNINFO] = { 0 };
+
+	if (!config->monitorDisabled)
+	{
+		if (!monitor_get_most_advanced_standby(monitor,
+											   config->formation,
+											   groupId,
+											   &upstreamNode))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+	}
+	else
+	{
+		upstreamNode = keeper->otherNodes.nodes[0];
+	}
+
+	/*
+	 * Postgres 10 does not have pg_replication_slot_advance(), so we don't
+	 * support replication slots on standby nodes there.
+	 */
+	if (pgSetup->control.pg_control_version >= 1100)
+	{
+		strlcpy(slotName, config->replication_slot_name, MAXCONNINFO);
+	}
+
+	if (!standby_init_replication_source(postgres,
+										 &upstreamNode,
+										 PG_AUTOCTL_REPLICA_USERNAME,
+										 config->replication_password,
+										 slotName,
+										 config->maximum_backup_rate,
+										 config->backupDirectory,
+										 upstreamNode.lsn,
+										 config->pgSetup.ssl,
+										 keeper->state.current_node_id))
+	{
+		/* can't happen at the moment */
+		return false;
+	}
+
+	if (!standby_fetch_missing_wal(postgres))
```

Comment:
Given that we can have all standbys async, and in theory `standby_fetch_missing_wal ` might take really long. Do you think we need a timeout guard/GUC against potentially very long times to catch up?

I can see that it might complicate the code a bit (as the same node would likely to be re-elected). Just wanted to bring up to get your thoughts on this, in case you think this might be a problem for some users or not.